### PR TITLE
Add migration idempotency checks 

### LINKED ISSUE #42: `on-runtime-upgrade` should enforce migrations are idempotent
See https://github.com/paritytech/polkadot-sdk/pull/1921#issue-1947857170

Should enforce by default, and allow skipping by specifying a CLI flag.

### PR DESCRIPTION
**Pull request by liamaharon**
Closes #42 

Checks that `on_runtime_upgrade` succeeds when called twice, and if it does succeed after being called twice, the storage state root did not change.

### CODE REVIEW

File: core/tests/execute_block.rs [hunk 1]
```diff
@@ -77,6 +77,9 @@ async fn execute_block_works() {
 
         // Assert that the block-execution process has executed a block.
         assert!(matched.is_ok());
+
+        let out = block_execution.wait_with_output().await.unwrap();
+        assert!(out.status.success());
```
Comment (pmikolajczyk41) [inline core/tests/execute_block.rs:82]: this was just missing, unrelated change, right?

File: core/tests/on_runtime_upgrade.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        println!("Checking idempotency execution issue can be ignored with an arg");
+        let child = on_runtime_upgrade(
+            snap_path.as_str(),
+            runtime_path.as_str(),
+            &["--no-idempotency-checks"],
+        );
+        let out = child.wait_with_output().await.unwrap();
+        assert!(out.status.success());
+
+        println!("Checking migration with idempotency state root issue fails");
```
Comment (pmikolajczyk41) [inline core/tests/on_runtime_upgrade.rs:93]: I suggest splitting this huge testcase into smaller functions and lifting printlns into testcase names. Hopefully this won't lead to a dramatic increase in a CI time

File: core/src/lib.rs [hunk 3]
```diff
@@ -152,23 +152,22 @@ impl TryFrom<Vec<u8>> for RefTimeInfo {
...
     data: &[u8],
     mut extensions: Extensions,
     maybe_export_proof: Option<PathBuf>,
-) -> sc_cli::Result<(OverlayedChanges<HashingFor<Block>>, StorageProof, Vec<u8>)> {
-    let mut changes = Default::default();
-    let backend = ext.backend.clone();
-    let runtime_code_backend = sp_state_machine::backend::BackendRuntimeCode::new(&backend);
-    let proving_backend = TrieBackendBuilder::wrap(&backend)
+) -> sc_cli::Result<(StorageProof, Vec<u8>)> {
+    let runtime_code_backend = sp_state_machine::backend::BackendRuntimeCode::new(&ext.backend);
```
Comment (pmikolajczyk41) [inline core/src/lib.rs:162]: are we sure that we don't have to clone backend?

File: core/tests/on_runtime_upgrade.rs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            snap_path.as_str(),
+            runtime_path.as_str(),
+            &["--no-weight-warnings"],
+        );
+        let out = child.wait_with_output().await.unwrap();
+        assert!(out.status.success());
+
+        println!("Checking migration with idempotency execution issue fails");
+        let runtime_path = format!(
+            "{}/tests/runtimes/bridge_hub_rococo_runtime_NOT_IDEMPOTENT_EXECUTION.compact.compressed.wasm",
```
Comment (pmikolajczyk41) [inline core/tests/on_runtime_upgrade.rs:77]: it'd be nice to put some brief info/instructions how did you get these blobs (maybe in `tests/README.md` or something like that)

File: core/src/commands/on_runtime_upgrade.rs [hunk 5]
```diff
@@ -120,43 +125,93 @@ where
...
+                "🔬 Running TryRuntime_on_runtime_upgrade again to check idempotency: {:?}",
+                command.checks
+            );
+            let (oc_pre_root, _) = overlayed_changes.storage_root(&ext.backend, ext.state_version);
+            match state_machine_call_with_proof::<Block, HostFns>(
+                &ext,
+                &mut overlayed_changes,
+                &executor,
+                "TryRuntime_on_runtime_upgrade",
+                command.checks.encode().as_ref(),
```
Comment (pmikolajczyk41) [inline core/src/commands/on_runtime_upgrade.rs:159]: let's use literally same command arguments as above, i.e. `UpgradeCheckSelect::None.encode().as_ref()`

**Review by pmikolajczyk41 — APPROVED** [review]
cool!

File: core/src/commands/on_runtime_upgrade.rs [hunk 6]
```diff
@@ -120,43 +125,93 @@ where
...
+                Default::default(), // we don't really need any extensions here.
+                shared.export_proof.clone(),
+            ) {
+                Ok(_) => {
+                    // Exeuction was OK, check if the storage root changed.
```
Comment (ggwpez) [inline core/src/commands/on_runtime_upgrade.rs:164]:
```suggestion
                    // Execution was OK, check if the storage root changed.
```

File: core/src/commands/on_runtime_upgrade.rs [hunk 7]
```diff
@@ -120,43 +125,93 @@ where
...
+        true => {
+            log::info!("ℹ Skipping idempotency check");
+            true
+        }
+        false => {
+            log::info!(
+                "🔬 Running TryRuntime_on_runtime_upgrade again to check idempotency: {:?}",
+                command.checks
+            );
+            let (oc_pre_root, _) = overlayed_changes.storage_root(&ext.backend, ext.state_version);
```
[the remark this answers is not in the document]
Comment (liamaharon) [reply core/src/commands/on_runtime_upgrade.rs:153]: I originally tried that approach, but IIRC cmp is not implemented for overlay changes. 

This is very much fast enough though, and unlikely to have any corner cases missed :)

File: core/tests/on_runtime_upgrade.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            "{}/tests/runtimes/bridge_hub_rococo_runtime_OK.compact.compressed.wasm",
+            project_root
+        );
+        let child = on_runtime_upgrade(snap_path.as_str(), runtime_path.as_str(), &[]);
+        let out = child.wait_with_output().await.unwrap();
+        assert!(out.status.success());
+
+        println!("Checking migration with weight issue fails");
+        let runtime_path = format!(
+            "{}/tests/runtimes/bridge_hub_rococo_runtime_WEIGHT_ISSUE.compact.compressed.wasm",
```
[the remark this answers is not in the document]
Comment (liamaharon) [reply core/tests/on_runtime_upgrade.rs:59]: The files are only about 1MB in size, there aren't many and they will seldom change.

If they were larger or were likely to change often then I'd add LFS but I think it's overkill here.

File: core/src/commands/on_runtime_upgrade.rs (hunk 5 above)
Comment (liamaharon) [reply core/src/commands/on_runtime_upgrade.rs:159]: why? if the dev is running with pre/post checks, we want to check that `pre_upgrade` and `post_upgrade` don't fail the second time they are executed

Comment (pmikolajczyk41) [reply core/src/commands/on_runtime_upgrade.rs:159]: oh, sorry, had some blackout, I should have asked to use `command.checks.encode().as_ref()` above as well; you want to run pre/post checks during the actual migration

Comment (liamaharon) [reply core/src/commands/on_runtime_upgrade.rs:159]: no worries :)

**Comment by liamaharon** [comment]
> As a follow up: we can mock the host function `ext_logging_log_version_1` to silence the logs on the second run, or selectively set the target/level to make it less noisy.

Oh this will be very useful also for silencing logs when measuring weight. Thanks!
